sopel/coretasks.py

```diff
@@ -238,12 +249,15 @@ def track_modes(bot, trigger):
             arg = Identifier(arg)
             for mode in modes:
                 priv = bot.channels[channel].privileges.get(arg, 0)
-                # Throw an exception if the two privilege-tracking data
-                # structures get out of sync. That should never happen.
-                # This is a good place to check that bot.channels is doing
+                # Log a warning if the two privilege-tracking data structures
+                # get out of sync. That should never happen.
+                # This is a good place to verify that bot.channels is doing
                 # what it's supposed to do before ultimately removing the old,
                 # deprecated bot.privileges structure completely.
-                assert priv == bot.privileges[channel].get(arg, 0)
+                if priv != bot.privileges[channel].get(arg, 0):
+                    LOGGER.warning("Privilege tracking mismatch in {}! Please "
+                                   "share a raw log with Sopel's developers, if"
+                                   " available".format(channel))
```

Comment:
I need to play with the logger a bit, because I don't really know how it actually behaves yet, but this is what I _want_:

  - This warning goes to the log file AND terminal output (if there is a terminal)
  - The contents of both data structures goes ONLY to log file, as they could each contain hundreds of users

If there's no terminal, this should probably get sent to IRC (a debug/log channel or PM to Sopel's owner)—but at this point, I'm almost into "another PR" territory with that kind of complexity.

Reply:
No data structures; went with just the values for the nick being processed.